### PR TITLE
Migrate PyPI release to GitHub Actions

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

This PR migrate the PyPI release job from Travis CI to GitHub Actions.

References:

- https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.
- https://github.com/fatiando/boule/tree/master/.github/workflows

Address #566 

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Notes**

- You can write `/format` in the first line of a comment to lint the code automatically

### CODE REVIEW

File: .github/workflows/publish-to-pypi.yml
Comment (weiji14) [inline .github/workflows/publish-to-pypi.yml:26]:
```suggestion
    - name: Build source and wheel distributions
      run: |
        # Change the versioneer format to "pre" so that the commit hash isn't
        # included (PyPI doesn't allow it). Can't do this permanently because
        # we rely the hash to indicate to the tests that this is a local
        # verison instead of a published version.
        sed --in-place "s/pep440/pep440-pre/g" setup.cfg
        python setup.py sdist bdist_wheel
        echo ""
        echo "Generated files:"
        ls -lh dist/
```

See if this works, inspired by https://github.com/fatiando/pooch/pull/196, specifically the lines at https://github.com/fatiando/pooch/blob/a9564b13a95cb4521dcf5938cea5adfd2851b385/.github/workflows/deploy.yml#L51-L61

**Review by weiji14 — COMMENTED** [review]
I made a suggestion following @leouieda's PR at https://github.com/fatiando/pooch/pull/196. **However**, I'm not sure if I like this workaround, since we will have to switch from versioneer to setuptools_scm (#671) sooner or later.

**Comment by seisman** [comment]
See the "fake" release in the test pypi repository: https://test.pypi.org/project/pygmt/0.2.0.post0.dev52/#history

File: .github/workflows/publish-to-pypi.yml
Comment (seisman) [inline .github/workflows/publish-to-pypi.yml:8]: We don't want to trigger the workflow for pushes and pull requests.
```suggestion
  # Runs for pull requests for testing purposes
  #pull_request:
  #  branches:
  #    - master
```

File: .github/workflows/publish-to-pypi.yml
Comment (seisman) [inline .github/workflows/publish-to-pypi.yml:42]: We may want to skip this step, as it's not needed for real releases.

File: .github/workflows/publish-to-pypi.yml
Comment (seisman) [inline .github/workflows/publish-to-pypi.yml:55]: We may skip this step, as it's unnecessary to publish to test PyPI after we merge this PR into master.

File: .github/workflows/publish-to-pypi.yml
Comment (weiji14) [reply .github/workflows/publish-to-pypi.yml:42]: Can we ask @leouieda (who has more experience) to weigh in on this?

Comment (seisman) [reply .github/workflows/publish-to-pypi.yml:42]: I believe keeping it or not makes no difference to tagged releases, as the version strings would always be "vX.Y.Z" with no suffix string.

Comment (leouieda) [reply .github/workflows/publish-to-pypi.yml:42]: Yep, this step doesn't make a difference for the actual releases since the version number generated is the same. We just merged a similar workaround for setuptools_scm in [fatiando/boule#61](https://github.com/fatiando/boule/pull/61/files#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3R70)

File: .github/workflows/publish-to-pypi.yml
Comment (leouieda) [reply .github/workflows/publish-to-pypi.yml:55]: It's not necessary but it's nice to have this test. It avoids headaches when publishing a release and suddenly things fail because PyPI changed something.

**Comment by leouieda** [comment]
I would actually recommend looking at the workflows here instead: https://github.com/fatiando/boule/tree/master/.github/workflows We're also trying out a tool called [nox](https://nox.thea.codes/en/stable/). It's pretty nice to automate some tasks with Python instead of clunky Makefiles that don't work on Windows.

File: .github/workflows/publish-to-pypi.yml
Comment (seisman) [reply .github/workflows/publish-to-pypi.yml:55]: To have the test, I think I should keep the workaround `sed --in-place "s/pep440/pep440-pre/g" setup.cfg`, and trigger the workflow for pushes to master branch and tagged releases, right?

Comment (leouieda) [reply .github/workflows/publish-to-pypi.yml:55]: 👍

File: .github/workflows/publish-to-pypi.yml
Comment (seisman) [inline .github/workflows/publish-to-pypi.yml:16]:
```suggestion
  # Runs for pull requests should be disable unless for testing purposes
  #pull_request:
  #  branches:
  #    - master
```

Comment (seisman) [reply .github/workflows/publish-to-pypi.yml:16]: This PR is ready for review. 

I will disable builds for pull requests after this PR is approved. After that, this workflow will be triggered for pushed to master branch (to TestPyPI only) and tagged releases (both PyPI and TestPyPI).

File: .github/workflows/publish-to-pypi.yml
Comment (weiji14) [inline .github/workflows/publish-to-pypi.yml:43]:
```suggestion
        # we rely on the hash to tell the tests that this is a local version
        # instead of a published version.
```

**Review by weiji14 — COMMENTED** [review]
These changes will have to be documented under MAINTENANCE.md. Could you remove mention of Travis CI PyPI deployments at https://github.com/GenericMappingTools/pygmt/blob/v0.2.0/MAINTENANCE.md#travis-ci and place it under the Github Actions section at https://github.com/GenericMappingTools/pygmt/blob/v0.2.0/MAINTENANCE.md#github-actions instead please?

File: MAINTENANCE.md
Comment (weiji14) [inline MAINTENANCE.md:72]:
```suggestion
4. `publish-to-pypi.yml` (Publish wheels to PyPI and TestPyPI)

This workflow is ran to publish wheels to PyPI and TestPyPI (for testing only).
Archives will be pushed to TestPyPI on every commit to the *master* branch and
tagged releases, and to PyPI for tagged releases only.
```

**Review by weiji14 — COMMENTED** [review]
Getting there! Just one minor suggestion to change 'archive' to 'wheel'.

**Review by weiji14 — APPROVED** [review]
Cool, thanks for doing this @seisman, and to @leouieda for your experience on the fatiando side! This will really streamline our CI going forward :smile:

P.S. remember to disable the Pull Request trigger.

File: .github/workflows/publish-to-pypi.yml
Comment (weiji14) [reply .github/workflows/publish-to-pypi.yml:59]: The Publish to Test PyPI step will fail for the next few pushes to master (0.2.0.post0.dev52 onwards) until we release PyGMT v0.2.1 because it's trying to reupload with the same name. See e.g.
https://github.com/GenericMappingTools/pygmt/runs/1368511293?check_suite_focus=true#step:8:27

image

A little annoying, but we can safely ignore it for now :slightly_smiling_face:

Comment (seisman) [reply .github/workflows/publish-to-pypi.yml:59]: Feel free to delete some releases.

Comment (weiji14) [reply .github/workflows/publish-to-pypi.yml:59]: Hmm, I can't login to TestPyPI with my PyPI account, are they separate entities? Also seems like you're the only maintainer for PyGMT on TestPyPI.

Comment (seisman) [reply .github/workflows/publish-to-pypi.yml:59]: Yes they are separate. I will add you and leo as maintainers later.

**Pull request merged**
